platform_generator.R

```diff
@@ -156,68 +107,62 @@ for(j in 1:nrow(gate_db)){
         round(gate_downloads[(j-1)*3+k-1,]$Download_End)
     } 
     gate_downloads[(j-1)*3+k,]$Download_End = 
-      gate_downloads[(j-1)*3+k,]$Download_Start + round(gate_downloads[(j-1)*3+k,]$Time/1000)
+      gate_downloads[(j-1)*3+k,]$Download_Start + round(gate_downloads[(j-1)*3+k,]$Time)
   }
 }
 
-gate_downloads$Bandwidth <-gate_downloads$FileSize/(pmax(0.1,(gate_downloads$Time-1000)))
-gate_downloads$Time <- gate_downloads$Time/1000
-
 gate_downloads$concurrency <- 1
 
-# sampling 20 points to estimate the concurrency of each transfer
-n_point <- 20
+# for merge download trqnsfers, the concurrency is considered as 1
+merge_downloads$concurrency <- 1
+
+
+
+# divide each transfer into 50 intervals to estimate the nominal bandwidth (only for Gate downloads)
+n_interval <- 50
 for(i in 1:nrow(gate_downloads)){
-  tmp <- vector(mode="numeric", length=n_point)
-  step <- (gate_downloads[i,]$Download_End - gate_downloads[i,]$Download_Start)/n_point
-  for(s in 1: n_point){
-    tmp[s] <- nrow(subset(gate_downloads, SE_SITE == gate_downloads[i,]$SE_SITE
+  conc <- vector(mode="numeric", length=n_interval)
+  step <- (gate_downloads[i,]$Download_End - gate_downloads[i,]$Download_Start)/n_interval
+  for(s in 1: n_interval){
+    conc[s] <- nrow(subset(gate_downloads, SE_SITE == gate_downloads[i,]$SE_SITE
                                          &FileSize == gate_downloads[i,]$FileSize
                                          &Download_Start <= (step*(s-1)+gate_downloads[i,]$Download_Start)
                                          &Download_End >= (step*(s-1)+gate_downloads[i,]$Download_Start)))
   }
-  # take the mean of concurrency of all sampling points
-  if(mean(tmp) > 1){
-    gate_downloads[i,]$concurrency <- mean(tmp)
-  }
-}
 
-## function to find the concurrency of a link
-## if the max is given by several transfers, the maximum concurrency is taken for this link.
-Concurrency_link <- function(Bandwidth, concurrency){
-  pos <- which(Bandwidth == max(Bandwidth))
-  con <- concurrency[pos[1]]
-  if(length(pos)>1){
-    for(i in 2:length(pos)){
-      if(concurrency[pos[i]]> con){
-        con <- concurrency[pos[i]]
-      }
-    }
-  }
-  return (con)
+  tmp <- 1/conc
+  gate_downloads[i,]$concurrency <- n_interval/sum(tmp)
 }
 
+gate_downloads$Bandwidth_improved <- gate_downloads$Bandwidth * gate_downloads$concurrency
+merge_downloads$Bandwidth_improved <- merge_downloads$Bandwidth * merge_downloads$concurrency
+
+
+df_download <- rbind(gate_downloads, merge_downloads)
```

Comment:
Is it really necessary to distinguish gate and merge downloads? What is the reason for that?